src/libvfcfuncinstrument/libVFCFuncInstrument.cpp

```diff
@@ -0,0 +1,416 @@
+      AU.addRequired<TargetLibraryInfoWrapperPass>();
+    }
+
+    // Fill have_double and have_float with true if the call_inst pi use at least 
+    // of the managed types
+    void haveFloatingPointArithmetic( Instruction *pi, 
+                                      bool is_from_library, 
+                                      bool is_intrinsic, 
+                                      bool *have_float, 
+                                      bool *have_double,
+                                      Module &M)
+    {
+      Function *f = cast<CallInst>(pi)->getCalledFunction();
+      Type *FloatTy = Type::getFloatTy(M.getContext());
+      Type *DoubleTy = Type::getDoubleTy(M.getContext());
+      Type *FloatPtrTy = Type::getFloatPtrTy(M.getContext());
+      Type *DoublePtrTy = Type::getDoublePtrTy(M.getContext());
+      Type* ReturnTy = pi->getType();
+
+      // Test if return type of pi is float
+      (*have_float) = (ReturnTy == FloatTy || ReturnTy == FloatPtrTy);
+
+      // Test if return type of pi is double
+      (*have_double) = (ReturnTy == DoubleTy || ReturnTy == DoublePtrTy);
+
+      // Test if f treat floats point numbers
+      if (is_intrinsic || is_from_library){
+        // Loop over arguments types 
+        for (auto it = pi->op_begin(); it < pi->op_end()-1; it++){
+          if ((*it)->getType() == FloatTy || (*it)->getType() == FloatPtrTy)
+            (*have_float) = true;
+          if ((*it)->getType() == DoubleTy || (*it)->getType() == DoublePtrTy)
+            (*have_double) = true;
+        }
+      }else{
+        // Loop over each instruction of the function and test if one of them 
+        // use float or double
+        for (auto bbi = f->begin(); bbi != f->end(); bbi++){
+          for (auto ii = bbi->begin(); ii != bbi->end(); ii++){
+            Type *opType = ii->getOperand(0)->getType();
+            
+            if (opType->isVectorTy()){
+              VectorType *t = static_cast<VectorType *>(opType);
+              opType = t->getElementType();
+            }
+
+            if (opType == FloatTy || opType == FloatPtrTy)
+              (*have_float) = true;
+
+            if (opType == DoubleTy || opType == DoublePtrTy)
+              (*have_double) = true;
+          }
+                  
+          if ((*have_float) && (*have_double))
+            break;
+        }
+      }
+    }
+
+    // Demangling function 
+    std::string demangle(std::string src)
+    {
+      int status = 0;
+      char *demangled_name = NULL;
+      if( (demangled_name = abi::__cxa_demangle(src.c_str(), 0, 0, &status))){
+        src = demangled_name;
+        std::size_t first = src.find("(");
+        src = src.substr(0,first);
+      }
+      free(demangled_name);
+
+      return src;
+    }
+
+    virtual bool runOnModule(Module &M)
+    {
+      const TargetLibraryInfo *TLI = &getAnalysis<TargetLibraryInfoWrapperPass>().getTLI();
+      const DataLayout &DL = M.getDataLayout();
+
+      // Types
```

Comment:
Do this once in the class (avoid doing it for every function). Use const for variables that are not changing.

Reply:
Fixed